File: app/src/main/java/com/example/schedule/compose/repo/HomeworkRepo.java
/*
package com.example.schedule.compose.repo;

import android.annotation.SuppressLint;
import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import com.example.schedule.ScheduleDBHelper;
import com.example.schedule.entity.Flow;
import com.example.schedule.entity.Homework;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;

public class HomeworkRepo {
    private final ScheduleDBHelper dbHelper;
    private final FlowRepo flowRepo;

    public HomeworkRepo(Context context) {
        this.dbHelper = new ScheduleDBHelper(context);
        this.flowRepo = new FlowRepo(context);
    }

    public HomeworkRepo(ScheduleDBHelper dbHelper, FlowRepo flowRepo) {
        this.dbHelper = dbHelper;
        this.flowRepo = flowRepo;
    }

    public long add(
            long flow, String lessonName, String homework, LocalDate lessonDate, int lessonNum
    ) {
        SQLiteDatabase db = dbHelper.getWritableDatabase();

        ContentValues values = new ContentValues();
        values.put("flow", flow);
        values.put("lesson_name", lessonName);
        values.put("homework", homework);
        values.put(
                "lesson_date",
                lessonDate.format(DateTimeFormatter.ofPattern("yyyy-MM-dd"))
        );
        values.put("lesson_num", lessonNum);

        long id = db.insert("homework", null, values);
        db.close();
        return id;
    }

    public long add(
            int flowLvl, int course, int flow, int subgroup,
            String lessonName, String homework, LocalDate lessonDate, int lessonNum
    ) {
        long flowId = -1;
        Flow foundFlow = flowRepo.findByFlowLvlAndCourseAndFlowAndSubgroup(
                flowLvl, course, flow, subgroup
        );
        if (foundFlow != null) flowId = foundFlow.getId();
        else flowId = flowRepo.add(flowLvl, course, flow, subgroup);

        return add(flowId, lessonName, homework, lessonDate, lessonNum);
    }

    public int update(
            long flow, String lessonName, String homework, LocalDate lessonDate, int lessonNum
    ) {
        SQLiteDatabase db = dbHelper.getWritableDatabase();

        ContentValues values = new ContentValues();
        values.put("lesson_name", lessonName);
        values.put("homework", homework);
        
        String whereClause = "flow=? AND lesson_date=? AND lesson_num=?";
        String[] whereArgs = new String[] {
                String.valueOf(flow), lessonDate.format(DateTimeFormatter.ofPattern("yyyy-MM-dd")),
                String.valueOf(lessonNum)
        };

        int count = db.update("homework", values, whereClause, whereArgs);
        db.close();
        return count;
    }
    
    public int update(
            int flowLvl, int course, int flow, int subgroup,
            String lessonName, String homework, LocalDate lessonDate, int lessonNum
    ) {
        long flowId = -1;
        Flow foundFlow = flowRepo.findByFlowLvlAndCourseAndFlowAndSubgroup(
                flowLvl, course, flow, subgroup
        );
        if (foundFlow != null) flowId = foundFlow.getId();
        else flowId = flowRepo.add(flowLvl, course, flow, subgroup);

        return update(flowId, lessonName, homework, lessonDate, lessonNum);
    }
    
    public void addOrUpdate(
            long flow, String lessonName, String homework, LocalDate lessonDate, int lessonNum
    ) {
        Homework foundHomework = findByFlowAndLessonDateAndLessonNum(
                flow, lessonDate, lessonNum
        );
        if (foundHomework != null) update(flow, lessonName, homework, lessonDate, lessonNum);
        else add(flow, lessonName, homework, lessonDate, lessonNum);
    }
    
    public void addOrUpdate(
            int flowLvl, int course, int flow, int subgroup,
            String lessonName, String homework, LocalDate lessonDate, int lessonNum
    ) {
        long flowId = -1;
        Flow foundFlow = flowRepo.findByFlowLvlAndCourseAndFlowAndSubgroup(
                flowLvl, course, flow, subgroup
        );
        if (foundFlow != null) flowId = foundFlow.getId();
        else flowId = flowRepo.add(flowLvl, course, flow, subgroup);

        addOrUpdate(flowId, lessonName, homework, lessonDate, lessonNum);
    }

    @SuppressLint("Range")
    public Homework findByFlowAndLessonDateAndLessonNum(
            long flow, LocalDate lessonDate, int lessonNum
    ) {
        Homework result = null;
        SQLiteDatabase db = dbHelper.getReadableDatabase();

        String sql = "SELECT * FROM homework " +
                "WHERE flow=? AND lesson_date=? AND lesson_num=?;";
        String[] selectionArgs = new String[] {
                String.valueOf(flow), lessonDate.format(DateTimeFormatter.ofPattern("yyyy-MM-dd")),
                String.valueOf(lessonNum)
        };
        Cursor cursor = db.rawQuery(sql, selectionArgs);
        if (cursor.moveToFirst()) {
            long id = cursor.getLong(cursor.getColumnIndex("id"));
            String lessonName = cursor.getString(cursor.getColumnIndex("lesson_name"));
            String homework = cursor.getString(cursor.getColumnIndex("homework"));

            result = new Homework();
            result.setId(id);
            result.setFlow(flow);
            result.setLessonName(lessonName);
            result.setHomework(homework);
            result.setLessonDate(lessonDate);
            result.setLessonNum(lessonNum);
        }
        cursor.close();
        db.close();
        return result;
    }

    @SuppressLint("Range")
    public Homework findByFlowAndLessonDateAndLessonNum(
            int flowLvl, int course, int flow, int subgroup, LocalDate lessonDate, int lessonNum
    ) {
        Flow foundFlow = flowRepo.findByFlowLvlAndCourseAndFlowAndSubgroup(
                flowLvl, course, flow, subgroup
        );
        if (foundFlow == null) return null;

        long flowId = foundFlow.getId();

        return findByFlowAndLessonDateAndLessonNum(flowId, lessonDate, lessonNum);
    }

    @SuppressLint("Range")
    public List<Homework> findAllByFlow(long flow) {
        List<Homework> result = new ArrayList<>();
        SQLiteDatabase db = dbHelper.getReadableDatabase();

        String sql = "SELECT * FROM homework WHERE flow=?;";
        String[] selectionArgs = new String[] {String.valueOf(flow)};
        Cursor cursor = db.rawQuery(sql, selectionArgs);
        while (cursor.moveToNext()) {
            long id = cursor.getLong(cursor.getColumnIndex("id"));
            String lessonName = cursor.getString(cursor.getColumnIndex("lesson_name"));
            String homework = cursor.getString(cursor.getColumnIndex("homework"));
            LocalDate lessonDate = LocalDate.parse(
                    cursor.getString(cursor.getColumnIndex("lesson_date")),
                    DateTimeFormatter.ofPattern("yyyy-MM-dd")
            );
            int lessonNum = cursor.getInt(cursor.getColumnIndex("lesson_num"));

            Homework foundHomework = new Homework();
            foundHomework.setId(id);
            foundHomework.setFlow(flow);
            foundHomework.setLessonName(lessonName);
            foundHomework.setHomework(homework);
            foundHomework.setLessonDate(lessonDate);
            foundHomework.setLessonNum(lessonNum);

            result.add(foundHomework);
        }
        cursor.close();
        db.close();
        return result;
    }

    public List<Homework> findAllByFlow(int flowLvl, int course, int flow, int subgroup) {
        Flow foundFlow = flowRepo.findByFlowLvlAndCourseAndFlowAndSubgroup(
                flowLvl, course, flow, subgroup
        );
        if (foundFlow == null) return null;

        long flowId = foundFlow.getId();

        return findAllByFlow(flowId);
    }

    public int deleteByFlowAndLessonDateAndLessonNum(
            long flow, LocalDate lessonDate, int lessonNum
    ) {
        SQLiteDatabase db = dbHelper.getWritableDatabase();

        String whereClause = "flow=? AND lesson_date=? AND lesson_num=?";
        String[] whereArgs = new String[] {
                String.valueOf(flow), lessonDate.format(DateTimeFormatter.ofPattern("yyyy-MM-dd")),
                String.valueOf(lessonNum)
        };

        int count = db.delete("homework", whereClause, whereArgs);
        db.close();
        return count;
    }

    public int deleteByFlowAndLessonDateAndLessonNum(
            int flowLvl, int course, int flow, int subgroup, LocalDate lessonDate, int lessonNum
    ) {
        Flow foundFlow = flowRepo.findByFlowLvlAndCourseAndFlowAndSubgroup(
                flowLvl, course, flow, subgroup
        );
        if (foundFlow == null) return 0;

        long flowId = foundFlow.getId();

        return deleteByFlowAndLessonDateAndLessonNum(flowId, lessonDate, lessonNum);
    }

    public int deleteAllBeforeDate(LocalDate date) {
        SQLiteDatabase db = dbHelper.getWritableDatabase();

        String whereClause = "lesson_date<?";
        String[] whereArgs = new String[] {
                date.format(DateTimeFormatter.ofPattern("yyyy-MM-dd"))
        };

        int count = db.delete("homework", whereClause, whereArgs);
        db.close();
        return count;
    }
}
*/

File: app/src/main/java/com/example/schedule/compose/repo/HomeworkRepo.kt
package com.example.schedule.compose.repo

import android.annotation.SuppressLint
import android.content.ContentValues
import android.database.Cursor
import com.example.schedule.compose.entity.Flow
import com.example.schedule.compose.entity.Homework
import com.example.schedule.compose.entity.Subject
import java.time.LocalDate
import java.time.format.DateTimeFormatter

class HomeworkRepo(
    private val dbHelper: ScheduleDBHelper,
    private val flowRepo: FlowRepo,
    private val subjectRepo: SubjectRepo
) {
    fun add(
        educationLevel: Int,
        course: Int,
        group: Int,
        subgroup: Int,
        homework: String,
        lessonDate: LocalDate,
        lessonNum: Int,
        subject: String
    ): Long {
        val foundFlow = flowRepo.findByEducationLevelAndCourseAndGroupAndSubgroup(educationLevel, course, group, subgroup)
        val flowId = foundFlow?.id ?: flowRepo.add(educationLevel, course, group, subgroup)

        val foundSubject = subjectRepo.findBySubject(subject)
        val subjectId = foundSubject?.id ?: subjectRepo.add(subject)

        val values = ContentValues()
        values.put("flow", flowId)
        values.put("homework", homework)
        values.put("lesson_date", lessonDate.format(DateTimeFormatter.ofPattern("yyyy-MM-dd")))
        values.put("lesson_num", lessonNum)
        values.put("subject", subjectId)

        val db = dbHelper.writableDatabase
        val id = db.insert("homework", null, values)
        db.close()
        return id
    }

    fun update(
        educationLevel: Int,
        course: Int,
        group: Int,
        subgroup: Int,
        homework: String,
        lessonDate: LocalDate,
        lessonNum: Int,
        subject: String
    ): Int {
        val foundFlow = flowRepo.findByEducationLevelAndCourseAndGroupAndSubgroup(educationLevel, course, group, subgroup)
        val flowId = foundFlow?.id ?: flowRepo.add(educationLevel, course, group, subgroup)

        val foundSubject = subjectRepo.findBySubject(subject)
        val subjectId = foundSubject?.id ?: subjectRepo.add(subject)

        val values = ContentValues()
        values.put("homework", homework)
        values.put("subject", subjectId)

        val whereClause = "flow=? AND lesson_date=? AND lesson_num=?"
        val whereArgs = arrayOf(flowId.toString(), lessonDate.format(DateTimeFormatter.ofPattern("yyyy-MM-dd")), lessonNum.toString())

        val db = dbHelper.writableDatabase
        val count = db.update("homework", values, whereClause, whereArgs)
        db.close()
        return count
    }

    fun addOrUpdate(
        educationLevel: Int,
        course: Int,
        group: Int,
        subgroup: Int,
        homework: String,
        lessonDate: LocalDate,
        lessonNum: Int,
        subject: String
    ) {
        val foundHomework = findByFlowAndLessonDateAndLessonNum(educationLevel, course, group, subgroup, lessonDate, lessonNum)
        foundHomework?.let {
            update(educationLevel, course, group, subgroup, homework, lessonDate, lessonNum, subject)
        } ?: run {
            add(educationLevel, course, group, subgroup, homework, lessonDate, lessonNum, subject)
        }
    }

    @SuppressLint("Range")
    fun findByFlowAndLessonDateAndLessonNum(
        educationLevel: Int,
        course: Int,
        group: Int,
        subgroup: Int,
        lessonDate: LocalDate,
        lessonNum: Int
    ): Homework? {
        var result: Homework? = null

        val sql = """
            SELECT h.id AS homework_id, h.flow AS flow_id, h.subject AS subject_id, h.homework, s.subject
            FROM homework h
            JOIN flow f ON h.flow = f.id
            JOIN subject s ON h.subject = s.id
            WHERE education_level=? AND course=? AND _group=? AND subgroup=? AND lesson_date=? AND lesson_num=?
        """.trimIndent()
        val selectionArgs = arrayOf(educationLevel.toString(), course.toString(), group.toString(), subgroup.toString(), lessonDate.format(DateTimeFormatter.ofPattern("yyyy-MM-dd")), lessonNum.toString())
        val db = dbHelper.readableDatabase
        val cursor: Cursor = db.rawQuery(sql, selectionArgs)
        if (cursor.moveToFirst()) {
            val flowId = cursor.getLong(cursor.getColumnIndex("flow_id"))

            val subjectId = cursor.getLong(cursor.getColumnIndex("subject_id"))
            val subject = cursor.getString(cursor.getColumnIndex("subject"))

            val id = cursor.getLong(cursor.getColumnIndex("homework_id"))
            val homework = cursor.getString(cursor.getColumnIndex("homework"))

            result = Homework(
                id,
                Flow(flowId, educationLevel, course, group, subgroup),
                homework,
                lessonDate,
                lessonNum,
                Subject(subjectId, subject)
            )
        }
        cursor.close()
        db.close()
        return result
    }

    @SuppressLint("Range")
    fun findAllByFlow(
        educationLevel: Int,
        course: Int,
        group: Int,
        subgroup: Int
    ): List<Homework> {
        val result = mutableListOf<Homework>()

        val sql = """
            SELECT h.id AS homework_id, h.flow AS flow_id, h.subject AS subject_id,
                h.homework, h.lesson_date, h.lesson_num, s.subject
            FROM homework h
            JOIN flow f ON h.flow = f.id
            JOIN subject s ON h.subject = s.id
            WHERE education_level=? AND course=? AND _group=? AND subgroup=?
            ORDER BY lesson_date, lesson_num
        """.trimIndent()
        val selectionArgs = arrayOf(educationLevel.toString(), course.toString(), group.toString(), subgroup.toString())
        val db = dbHelper.readableDatabase
        val cursor: Cursor = db.rawQuery(sql, selectionArgs)
        while (cursor.moveToNext()) {
            val flowId = cursor.getLong(cursor.getColumnIndex("flow_id"))

            val subjectId = cursor.getLong(cursor.getColumnIndex("subject_id"))
            val subject = cursor.getString(cursor.getColumnIndex("subject"))

            val id = cursor.getLong(cursor.getColumnIndex("homework_id"))
            val homework = cursor.getString(cursor.getColumnIndex("homework"))
            val lessonDate = LocalDate.parse(cursor.getString(cursor.getColumnIndex("lesson_date")), DateTimeFormatter.ofPattern("yyyy-MM-dd"))
            val lessonNum = cursor.getInt(cursor.getColumnIndex("lesson_num"))

            result.add(Homework(
                id,
                Flow(flowId, educationLevel, course, group, subgroup),
                homework,
                lessonDate,
                lessonNum,
                Subject(subjectId, subject)
            ))
        }
        cursor.close()
        db.close()
        return result
    }

    fun deleteByFlowAndLessonDateAndLessonNum(
        educationLevel: Int,
        course: Int,
        group: Int,
        subgroup: Int,
        lessonDate: LocalDate,
        lessonNum: Int
    ): Int {
        val foundFlow: Flow = flowRepo.findByEducationLevelAndCourseAndGroupAndSubgroup(educationLevel, course, group, subgroup) ?: return 0

        val whereClause = "flow=? AND lesson_date=? AND lesson_num=?"
        val whereArgs = arrayOf(foundFlow.id.toString(), lessonDate.format(DateTimeFormatter.ofPattern("yyyy-MM-dd")), lessonNum.toString())

        val db = dbHelper.writableDatabase
        val count = db.delete("homework", whereClause, whereArgs)
        db.close()
        return count
    }

    fun deleteAllBeforeDate(
        date: LocalDate
    ): Int {
        val whereClause = "lesson_date<?"
        val whereArgs = arrayOf(date.format(DateTimeFormatter.ofPattern("yyyy-MM-dd")))

        val db = dbHelper.writableDatabase
        val count = db.delete("homework", whereClause, whereArgs)
        db.close()
        return count
    }
}
